node/core/candidate-validation/src/lib.rs

```
@@ -0,0 +1,913 @@
+	ValidationCode, OmittedValidationData, PoV, CandidateDescriptor, LocalValidationData,
+	GlobalValidationData, OccupiedCoreAssumption, Hash, validation_data_hash,
+};
+use polkadot_parachain::wasm_executor::{self, ValidationPool, ExecutionMode};
+use polkadot_parachain::primitives::{ValidationResult as WasmValidationResult, ValidationParams};
+
+use parity_scale_codec::Encode;
+use sp_core::traits::SpawnNamed;
+
+use futures::channel::oneshot;
+use futures::prelude::*;
+
+use std::sync::Arc;
+
+/// The candidate validation subsystem.
+pub struct CandidateValidationSubsystem<S>(S);
+
+impl<S> CandidateValidationSubsystem<S> {
+	/// Create a new `CandidateValidationSubsystem` with the given task spawner.
+	pub fn new(spawn: S) -> Self {
+		CandidateValidationSubsystem(spawn)
+	}
+}
+
+impl<S, C> Subsystem<C> for CandidateValidationSubsystem<S> where
+	C: SubsystemContext<Message = CandidateValidationMessage>,
+	S: SpawnNamed + Clone + 'static,
+{
+	fn start(self, ctx: C) -> SpawnedSubsystem {
+		SpawnedSubsystem {
+			name: "candidate-validation-subsystem",
+			future: run(ctx, self.0).map(|_| ()).boxed(),
+		}
+	}
+}
+
+async fn run(
+	mut ctx: impl SubsystemContext<Message = CandidateValidationMessage>,
+	spawn: impl SpawnNamed + Clone + 'static,
+)
+	-> SubsystemResult<()>
+{
+	let pool = ValidationPool::new();
+
+	loop {
+		match ctx.recv().await? {
+			FromOverseer::Signal(OverseerSignal::ActiveLeaves(_)) => {}
+			FromOverseer::Signal(OverseerSignal::BlockFinalized(_)) => {}
+			FromOverseer::Signal(OverseerSignal::Conclude) => return Ok(()),
+			FromOverseer::Communication { msg } => match msg {
+				CandidateValidationMessage::ValidateFromChainState(
+					descriptor,
+					pov,
+					response_sender,
+				) => {
+					let res = spawn_validate_from_chain_state(
+						&mut ctx,
+						Some(pool.clone()),
+						descriptor,
+						pov,
+						spawn.clone(),
+					).await;
+
+					match res {
+						Ok(x) => { let _ = response_sender.send(x); }
```

Should we warn if sending failed rather than silently dropping the result

---

Sure. We might get some spurious warnings but probably OK.